content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,1460 @@
+
+1. In the `library` folder, add a new file `4_helpers.js`. Then add the following helper function to that file:
+
+{{ include_code("_code-samples/build-a-wallet/desktop-js/library/4_helpers.js", language="js") }}
+
+2. Now, in `index.js`, require the new helper function at the bottom of the import section like so:
+
+```javascript
+const { prepareReserve, prepareAccountData, prepareLedgerData} = require('./library/3_helpers')
+const { prepareTxData } = require('./library/4_helpers')
+```
+
+3. In `index.js`, update the listener function subscribed to the `transaction` event by adding the following snippet:
+
+```javascript
+// Wait for transaction on subscribed account and re-request account data
+client.on("transaction", async (transaction) => {
+  // Reference: https://xrpl.org/account_info.html
+  const accountInfoRequest = {
+    "command": "account_info",
+    "account": address,
+    "ledger_index": transaction.ledger_index
+  }
+
+  const accountInfoResponse = await client.request(accountInfoRequest)
+  const accountData = prepareAccountData(accountInfoResponse.result.account_data, reserve)
+  appWindow.webContents.send('update-account-data', accountData)
+
+  // Step 4 code additions - start
+  const transactions = prepareTxData([{tx: transaction.transaction}])
+  appWindow.webContents.send('update-transaction-data', transactions)
+  // Step 4 code additions - end
+  
+})
+```
+
+4. In `view/preload.js`, add the following code at the bottom of `exposeInMainWorld()`:
+
+```javascript
+contextBridge.exposeInMainWorld('electronAPI', {
+  onUpdateLedgerData: (callback) => {
+    ipcRenderer.on('update-ledger-data', callback)
+  },
+  onEnterAccountAddress: (address) => {
+    ipcRenderer.send('address-entered', address)
+  },
+  onUpdateAccountData: (callback) => {
+    ipcRenderer.on('update-account-data', callback)
+  },
+
+  // Step 4 code additions - start
+  onUpdateTransactionData: (callback) => {
+    ipcRenderer.on('update-transaction-data', callback)
+  }
+  // Step 4 code additions - end
+})
+```
+
+5. Modify `view/template.html` by adding a new fieldset below the ones that are already there:
+
+```html
+        ...
+        Close Time: <strong id="ledger-close-time"></strong><br/>
+    </fieldset>
+
+    <!-- Step 4 code additions - start -->
+    <fieldset>
+        <legend>Transactions:</legend>
+        <table id="tx-table">
+            <thead>
+            <tr>
+                <th>Confirmed</th>
+                <th>Type</th>
+                <th>From</th>
+                <th>To</th>
+                <th>Value Delivered</th>
+                <th>Hash</th>
+            </tr>
+            </thead>
+            <tbody></tbody>
+        </table>
+    </fieldset>
+    <!-- Step 4 code additions - end -->
+
+    <dialog id="account-address-dialog">
+        <form method="dialog">
+```
+
+The table here will be filled dynamically with the accounts transactions.
+
+6. Add the following code at the bottom of `view/renderer.js`:
```

Comment:
This is missing a step 7 to add the custom code to the beginning of `renderer.js` from `4-tx-history/view/renderer.js`

The missing code snippet looks like:
```
// Step 4 code additions - start
document.addEventListener('DOMContentLoaded', openAccountAddressDialog);

function openAccountAddressDialog(){
    const accountAddressDialog = document.getElementById('account-address-dialog');
    const accountAddressInput = accountAddressDialog.querySelector('input');
    const submitButton = accountAddressDialog.querySelector('button[type="submit"]');

    submitButton.addEventListener('click', () => {
        const address = accountAddressInput.value;
        window.electronAPI.onEnterAccountAddress(address)
        accountAddressDialog.close()
    });

    accountAddressDialog.showModal()
}
// Step 4 code additions - end
```

(Without it, the transactions relating to the account don't populate)

Reply:
These lines should be there from Step 3. From the instructions it says to replace the file in Step 3 as the changes are so grave - but there were misleading Sections comments in Step 4, I've fixed that. Will leave it open for review, though.

Reply:
Looks good - I think I missed it on my last pass through :)